### PR TITLE
Reduce allocations for map compaction: Overall 264 GB -> 126 GB

### PR DESCRIPTION
**Pull request by etiennedi**
## Status

- Please wait with merging until the `v1.12.2` release is out, so we can include this in `v1.13.0`
- All existing tests are passing as well as new tests were added. 
- Has not yet been run in the chaos pipeline

## Motivation
While fixing #1903 we created a new test setup that only used the inverted index (LSM store), but no vector index. While running this setup it became quite obvious that this kind of usage puts a ton of load on the GC, see #1908 for details.

## Changes
* The scope for this PR is only the Map Compactor. No changes were made to other Compactor types or other areas.
* One by one (you can browse the commits from oldest to newest) the worst offender according to the alloc profile was tackled.
* For each element that made a lot of allocations it was rewritten in a way that memory allocated "in the previous round" can be reused in the next.
* By generally choosing an allocation overhead of 25% we can completely eliminate future allocations if they fit. This seems to be the case for the test setup used. This may not apply to other setups, we may want to think about making this value configurable. For now I don't see any need for this.

### Setup
| Field | Value |
| --- | --- |
| Test Environment | MBP 2021 M1 Max |
| Compiler |`go version go1.17.6 darwin/arm64` |
| Import script | [`apps/importer-no-vector-index`](https://github.com/semi-technologies/weaviate-chaos-engineering/tree/main/apps/importer-no-vector-index) |
| Import parameters | `SHARDS=1 BATCH_SIZE=128 SIZE=200000 ORIGIN=http://localhost:8080 go run .` |

### Results

| Area | Value before | Value after | Improvement | Screenshots |
| --- | --- | --- | --- | --- |
| Overall allocations | 264 GB | 126 GB | 109 % | Before:<br>Screen Shot 2022-04-12 at 1 09 30 PMAfter:<br>Screen Shot 2022-04-12 at 1 09 42 PM |
| MapCompactor allocations | 144GB | < 10 GB (*does not show up on the profile at all anymore*) | 1340% | Before: <br>Screen Shot 2022-04-12 at 1 08 40 PM After: <br>Screen Shot 2022-04-12 at 1 10 10 PM |
| Import time | 176s | 149s | 18% | Before: <br>Screen Shot 2022-04-12 at 1 27 35 PMAfter: <br>Screen Shot 2022-04-12 at 1 24 14 PM |

## Next steps

Looking at the current alloc profile, it seems the next worst offenders are the `inverted.Analyzer`, the Memtable itself (adding), and the Memtable Flush cycle
profile014

### CODE REVIEW

File: adapters/repos/db/lsmkv/segment_serialization.go [hunk 1]
```diff
@@ -621,3 +622,79 @@ func ParseCollectionNode(r io.Reader) (segmentCollectionNode, error) {
...
+
+	node.offset = offset
+	return nil
+}
+
+func resizeValuesOfCollectionNode(node *segmentCollectionNode, size uint64) {
+	if cap(node.values) >= int(size) {
+		node.values = node.values[:size]
+	} else {
+		fmt.Println(size)
```
Comment (antas-marcin) [inline adapters/repos/db/lsmkv/segment_serialization.go:680]: Forgot to remove :)

Comment (etiennedi) [reply adapters/repos/db/lsmkv/segment_serialization.go:680]: Oh, good catch. There may be other places. It's still in Draft mode :-) I also saw that commited a local change to the `.dockerignore` ;-)

File: adapters/repos/db/lsmkv/compactor_map.go [hunk 2]
```diff
@@ -104,6 +104,9 @@ func (c *compactorMap) writeKeys() ([]keyIndex, error) {
 	offset := SegmentHeaderSize
 
 	var kis []keyIndex
+	pairs := newReusableMapPairs()
+	me := newMapEncoder()
+	ssm := newSortedMapMerger()
 
 	for {
 		// TODO: each iteration makes a massive amount of allocations, this could
```
Comment (parkerduckworth) [inline adapters/repos/db/lsmkv/compactor_map.go:112]: is this comment still relevant? 🙂

File: adapters/repos/db/lsmkv/segment_serialization.go [hunk 3]
```diff
@@ -621,3 +622,79 @@ func ParseCollectionNode(r io.Reader) (segmentCollectionNode, error) {
...
+// that memory may be shared between the two. As a result, the caller must make
+// sure that they do not modify "in" while "node" is still in use. A safer
+// alternative is to use ParseCollectionNode.
+//
+// The primary intention of this function is to provide a way to reuse buffers
+// when the lifetime is controlled tightly, for example in cursors used within
+// compactions. Use at your own risk!
+//
+// If the buffers of the provided node have enough capacity they will be
+// reused. Only if the capacity is not enogh, will an allocation occur. This
```
Comment (parkerduckworth) [inline adapters/repos/db/lsmkv/segment_serialization.go:637]: enogh -> enough

File: adapters/repos/db/lsmkv/segment_serialization.go [hunk 4]
```diff
@@ -621,3 +622,79 @@ func ParseCollectionNode(r io.Reader) (segmentCollectionNode, error) {
...
+	node.offset = offset
+	return nil
+}
+
+func resizeValuesOfCollectionNode(node *segmentCollectionNode, size uint64) {
+	if cap(node.values) >= int(size) {
+		node.values = node.values[:size]
+	} else {
+		fmt.Println(size)
+		node.values = make([]value, size, int(float64(size)*1.25))
```
Comment (parkerduckworth) [inline adapters/repos/db/lsmkv/segment_serialization.go:681]: Is this `*1.25` due to the same 25% overhead heuristic as mentioned above?

Comment (parkerduckworth) [reply adapters/repos/db/lsmkv/segment_serialization.go:681]: I see it several times below as well, like in `(rmp *reusableMapPairs) ResizeLeft`, `resizeValueOfCollectionNodeAtPos`, `resizeKeyOfCollectionNode`, etc. Assuming all for same reason?

File: adapters/repos/db/lsmkv/strategies_map_sorted_merger.go [hunk 5]
```diff
@@ -92,6 +111,42 @@ func (s *sortedMapMerger) init(segments [][]MapPair) error {
 	return nil
 }
 
+// reset can be manually called if sharing allocated state is desired, such as
+// with .doKeepTombstonesReusable()
+func (s *sortedMapMerger) reset(segments [][]MapPair) error {
```
Comment (parkerduckworth) [inline adapters/repos/db/lsmkv/strategies_map_sorted_merger.go:116]: this is very easy to understand with these comments

**Review by parkerduckworth — APPROVED** [review]

File: adapters/repos/db/lsmkv/segment_serialization.go (hunk 4 above)
Comment (etiennedi) [reply adapters/repos/db/lsmkv/segment_serialization.go:681]: Correct, I'll add a one-line comment each time this occurs.

File: adapters/repos/db/lsmkv/strategies_map_sorted_merger.go (hunk 5 above)
Comment (etiennedi) [reply adapters/repos/db/lsmkv/strategies_map_sorted_merger.go:116]: Thanks 😍

**Review by aliszka — APPROVED** [review]

**Pull request merged**